store/siren-action-behavior.js

```diff
@@ -103,11 +103,19 @@ D2L.PolymerBehaviors.Siren.SirenActionBehaviorImpl = {
 			.then(function(resp) {
 				if (!resp.ok) {
 					var errMsg = resp.statusText + ' response executing ' + opts.method + ' on ' + href + '.';
-					return resp.json().then(function(data) {
-						throw { json: data, message: errMsg };
-					}, function(data) {
-						throw { string: data, message: errMsg };
-					});
+					return resp.text()
+						.then(function(text) {
+							try {
+								return JSON.parse(text);
+							} catch (e) {
+								return text;
+							}
+						})
+						.then(function(data) {
+							throw { json: data, message: errMsg };
```

Comment:
Wouldn't this give a `{ json: 'some string' }` due to the above block simply returning a string?

Maybe create the error object in the above block instead?

Reply:
I don't follow you. The string that is in the `json` field is exactly the data that we want to surface. The whole point of this PR is that we don't receive that data at all, because of some format issues. To me creating an error object is secondary because receiving the error message for is us is really important.

Maybe I'm misunderstanding your concern?

Reply:
If
https://github.com/Brightspace/polymer-siren-behaviors/pull/65/files/434b169f98b4e6ac5f6c94fb5b3d9482b0642e26#diff-b93d846bf098fc66ad5f4bd5ffca5af6R108-R110
throws, then
https://github.com/Brightspace/polymer-siren-behaviors/pull/65/files/434b169f98b4e6ac5f6c94fb5b3d9482b0642e26#diff-b93d846bf098fc66ad5f4bd5ffca5af6R111
is executed, which never throws.

Therefore, https://github.com/Brightspace/polymer-siren-behaviors/pull/65/files/434b169f98b4e6ac5f6c94fb5b3d9482b0642e26#diff-b93d846bf098fc66ad5f4bd5ffca5af6R111 will never be run.

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then#Parameters `onRejected`

Reply:
yeahhhhhh the first part is by design.

the reason why i kept that function at the bottom, is because I don't understand it, so I thought that there might be a use case where it would actually get called.

If this piece of code never gets run, I'm very happy to get rid of it ✂️

Reply:
I would also make the `json` field so that it isn't always used by checking the type of `data`

Reply:
I tried different things to get the type of data in the response. Information in the response doesn't really change based on whether the response is "json" or just text. I found that the only way to find out, is to try parsing the data and hope it doesn't blow up in our face.

Reply:
```javascript
if (typeof data === 'string') {
return { string: data, message: ... };
} else {
return { json: data, message: ... };
}
```

Reply:
I feel the fact that we have a discrepancy between the types of output is an inconvenience. This is because there is no meaningful reason why we sometimes our errors are in json and sometimes they are text. I feel that if we return two types of errors, one json and one text, it implies to the consumer that there might be an important difference between them, which there isn't.

So for that reason, I think what is important is the actual error message so that we can pick it up in our logs, and I feel that this library should try simplify error message. If we have two types, the consumer then has to check which type error is it , JSON? or Text?, but then in my experience there is no real difference in the content of the messages. 

Ideally, I wish all error messages should be in json, for HM purposes.

This is just my opinion.

Reply:
I was thinking along the lines of consistency with the existing behavior. And I agree all errors should be JSON, but they aren't here, and so should be handled differently by the consumer.

The error message isn't always very useful. It just gives the error given by the fetch API. Like `Not Found` or `Network Error`. It can be useful to get the server response text, so should still be logged. But I guess it doesn't matter if it is json or text.

Maybe something like this then?

```javascript
return { data, message: ... };
```